rfcs/2021-09-21-9292-throttle-transform.md

```diff
@@ -0,0 +1,149 @@
+
+## Context
+
+- [#258](https://github.com/vectordotdev/vector/issues/258)
+
+## Scope
+
+### In scope
+
+- Dropping events (logs and metrics) to rate limit an event stream
+- Rate limit by both bytes (unserialized) and events
+- Exclude events with VRL conditions
+- Specify buckets based on the value of a key in the event
+
+### Out of scope
+
+- Apply backpressure as a form of rate limiting
+
+## Pain
+
+- Users lack the necessary tooling to control throughput which can cause excessive costs and negatively impact downstream services due to increased load
+- Admins cannot set quotas for users/usergroups utilizing Vector
+
+## Proposal
+
+### User Experience
+
+The `throttle` transform can be used to rate limit specific subsets of your event stream to limit load on downstream services or to enforce quotas on users.
+You can enforce rate limits on events or their raw size, as well as excluding events based on a VRL condition to avoid dropping critical logs. Rate limits
+can be applied globally across all logs or by specifying a key to create buckets of events to rate limit more granularly.
+
+### Implementation
+
+The `throttle` transform will leverage the existing [Governor](https://docs.rs/governor/0.3.2/governor/index.html) crate. We will expose the bytes or events
+per second as using facing configuration (or expand to more time frames) and translate that into a `governor::Quota` for internal use. We should not expose
+the `Quota` directly to users to ensure ease of configuration for end users.
+
+Config:
+
+```rust
+pub struct ThrottleConfig {
+    events_per_second: u32,
+    bytes_per_second: u32,
+    key_field: Option<String>,
+    exclude: Option<AnyCondition>,
+}
+```
+
+TaskTransform:
+
+```rust
+impl TaskTransform for Throttle {
+    fn transform(
+        self: Box<Self>,
+        mut input_rx: Pin<Box<dyn Stream<Item = Event> + Send>>,
+    ) -> Pin<Box<dyn Stream<Item = Event> + Send>>
+    where
+        Self: 'static,
+    {
+        let lim = RateLimiter::keyed(Quota::per_second(self.events_per_second));
+
+        let mut flush_stream = tokio::time::interval(Duration::from_millis(1000));
+
+        Box::pin(
+            stream! {
+              loop {
+                let mut output = Vec::new();
+                let done = tokio::select! {
+                    _ = flush_stream.tick() => {
+                        false
+                    }
+                    maybe_event = input_rx.next() => {
+                        match maybe_event {
+                            None => true,
+                            Some(event) => {
+			    	if let Some(condition) = self.exclude_as_ref() {
+				    if condition.check(&event) {
+				        output.push(event);
+					false
+				    }
+				}
+
+				let value = self
+				    .key_field
+				    .as_ref()
+				    .and_then(|key_field| event.get(key_field))
+				    .map(|v| v.to_string_lossy());
+
+                                match lim.check_key_n(value, size) {
```

Comment:
nit: probably good to `rustfmt` these implementation snippets.

Reply:
_waves hands_ pseudocode! Some of it was copied from the `sample` transform, and some of the bits aren't yet working from my current spike :)

Reply:
Hah, yeah. Well, hand-crafted `rustfmt` then. At least _some_ effort to avoid the weird tab/spaces difference would be good 😄

Reply:
Definitely will do - I hadn't realized it got so hideous 😆 it looked fine in my editor at the time!